lib/Transforms/Scalar/DxilConditionalMem2Reg.cpp

```diff
@@ -0,0 +1,246 @@
+//===- DxilConditionalMem2Reg.cpp - Mem2Reg that selectively promotes Allocas ----===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Pass.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/Transforms/Scalar.h"
+#include "llvm/Transforms/Utils/PromoteMemToReg.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/IR/PredIteratorCache.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+
+#include "dxc/DXIL/DxilUtil.h"
+#include "dxc/HLSL/HLModule.h"
+#include "llvm/Analysis/DxilValueCache.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+static bool ContainsFloatingPointType(Type *Ty) {
+  if (Ty->isFloatingPointTy()) {
+    return true;
+  }
+  else if (Ty->isArrayTy()) {
+    return ContainsFloatingPointType(Ty->getArrayElementType());
+  }
+  else if (Ty->isVectorTy()) {
+    return ContainsFloatingPointType(Ty->getVectorElementType());
+  }
+  else if (Ty->isStructTy()) {
+    for (unsigned i = 0, NumStructElms = Ty->getStructNumElements(); i < NumStructElms; i++) {
+      if (ContainsFloatingPointType(Ty->getStructElementType(i)))
+        return true;
+    }
+  }
+  return false;
+}
+
+static bool Mem2Reg(Function &F, DominatorTree &DT, AssumptionCache &AC) {
+  BasicBlock &BB = F.getEntryBlock();  // Get the entry node for the function
+  bool Changed  = false;
+  std::vector<AllocaInst*> Allocas;
+  while (1) {
+    Allocas.clear();
+
+    // Find allocas that are safe to promote, by looking at all instructions in
+    // the entry node
+    for (BasicBlock::iterator I = BB.begin(), E = --BB.end(); I != E; ++I)
+      if (AllocaInst *AI = dyn_cast<AllocaInst>(I))       // Is it an alloca?
+        if (isAllocaPromotable(AI) &&
+          (!HLModule::HasPreciseAttributeWithMetadata(AI) || !ContainsFloatingPointType(AI->getAllocatedType())))
+          Allocas.push_back(AI);
+
+    if (Allocas.empty()) break;
+
+    PromoteMemToReg(Allocas, DT, nullptr, &AC);
+    Changed = true;
+  }
+
+  return Changed;
+}
+
+class DxilConditionalMem2Reg : public FunctionPass {
```

Comment:
You removed the old comments here. Could you add new ones? Is it that this isn't just done to facilitate loopunrolling anymore? The selectivity it practiced seems unchanged. Information about how it chooses what to promote would be helpful